lib/kernel/src/erl_erlang_errors.erl

```diff
@@ -0,0 +1,811 @@
+%%
+%% %CopyrightBegin%
+%%
+%% Copyright Ericsson AB 2020. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+
+-module(erl_erlang_errors).
+-export([format_error/4]).
+
+-spec format_error(Function, Args, ExceptionReason, ErrorInfoMap) -> ErrorMap when
+      Function :: atom(),
+      Args :: [term()],
+      ExceptionReason :: term(),
+      ErrorInfoMap :: #{},
+      ErrorMap :: #{pos_integer() => unicode:chardata()}.
```

Comment:
`pos_integer()` might not be enough in general. Take the example of `ssl:connect/3`. It has a couple arguments followed by `Opts`. There may be tons of options. Being limited strictly to argument position and text would potentially leave you with a wall of text full of errors.

Could the error map also allow something like `#{{pos_integer(), atom()} => <<"unknown protocol nottls">>}` for these cases? Then it can be formatted as `argument 3 (protocol): unknown protocol nottls`.

The intent is about separating independent errors, although a library could also use this to describe what the argument is beyond its position.

Reply:
I don't see anything in `ssl:connect/3` documentation that it would raise exceptions. This EEP is about exceptions, not errors returned from a function.

Reply:
I only used `ssl:connect/3` as an example of a function with lots of options, not necessarily as one that immediately applies to this. Functions like `term_to_binary` for example will only tell you `"invalid option in list"`. That's fine because the options list is small. But if a user library has many options, similar to what `ssl:connect/3` has, and triggers an exception, then knowing the exact location would be helpful.

The user's `format_error` function could very well put the exact location directly in the message, but the message might grow large if there are many different errors in the options. Having something like `{pos_integer(), atom()}` or even `{pos_integer(), any()}` allowed as key would instead result in multiple errors for the same argument which is more readable.

We could also argue that functions such as `ssl:connect/3`, like many other functions I've written, should not return an error tuple when the arguments are wrong, but that they only do so because the `error:badarg` exception was not helpful before this PR. This PR will most definitely impact the way we report errors to the user. I'd love to change functions like `gun:open` to trigger an exception if only I can properly inform the user of where the problem is, and that function has multiple levels of options. So output would ideally look something like:

```
argument 3 (http2_opts.bad): unknown option
argument 3 (keepalive): must be a timeout()
```

And it would also need to apply both to the shell and to logger reports.

Anyway this is not something that can't be improved later on so the stakes are low to get this right immediately.

Reply:
> We could also argue that functions such as `ssl:connect/3`, like many other functions I've written, should not return an error tuple when the arguments are wrong, but that they only do so because the error:badarg exception was not helpful before this PR.

A function that raises an exception does not have to raise a `badarg` exception. It could just as easily raise a `{badoption,DetailedReason}` exception. We are stuck with `badarg` for BIFs for backward compatibility reasons.

> Anyway this is not something that can't be improved later on so the stakes are low to get this right immediately.

Yes, that's why we have the maps so that this mechanism can be improved as needed in the future.

Reply:
A place where this may be useful for us would be for `badopt` in `erlang:spawn_request`.